### PR TITLE
C++:  More use of set literals.

### PR DESCRIPTION
**Pull request by geoffw0**
More use of set literals (`[, ]`), especially in models.  As well as being more readable and compact, for typical uses of `hasQualifiedName` we've found this often results in better performance.

[Looking at the diff] I'm suspicious the differences between places where we use `getName` vs `hasGlobalName` vs `hasGlobalOrStdName` are used is mostly arbitrary.  I'd prefer `hasGlobalOrStdName` where appropriate, but I've left that alone for now and stuck to adding the use of set literals.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll [hunk 1]
```diff
@@ -14,95 +14,49 @@ private class MallocAllocationFunction extends AllocationFunction {
...
+    or
+    hasGlobalName(["HeapAlloc"]) and // HeapAlloc(heap, flags, size)
+    sizeArg = 2
+    or
+    hasGlobalName([
+        "MmAllocatePagesForMdl", // MmAllocatePagesForMdl(minaddress, maxaddress, skip, size)
+        "MmAllocatePagesForMdlEx", // MmAllocatePagesForMdlEx(minaddress, maxaddress, skip, size, type, flags)
+        "MmAllocateNodePagesForMdlEx" // MmAllocateNodePagesForMdlEx(minaddress, maxaddress, skip, size, type, prefer, flags)
+      ]) and
+    sizeArg = 3
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll:59]: There are a few different ways I could have arranged this and a few similar predicates.  Please comment if you have a preferred design!

Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll:59]: I think the organization here is good, but given how many environments these classes cover, we should add some comments about where the functions come from.

**Review by jbj — COMMENTED** [review]
LGTM when Robert's comment and the failing tests have been addressed.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll:68]: Done.  Bit of a pain tracking the information down but its good to have it.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll [hunk 2]
```diff
@@ -14,19 +14,24 @@ private class MallocAllocationFunction extends AllocationFunction {
...
         "MmAllocateContiguousNodeMemory", // MmAllocateContiguousNodeMemory(size, minaddress, maxaddress, bound, flag, prefer)
         "MmAllocateContiguousMemorySpecifyCache", // MmAllocateContiguousMemorySpecifyCache(size, minaddress, maxaddress, bound, type)
         "MmAllocateContiguousMemorySpecifyCacheNode", // MmAllocateContiguousMemorySpecifyCacheNode(size, minaddress, maxaddress, bound, type, prefer)
         "MmAllocateNonCachedMemory", // MmAllocateNonCachedMemory(size)
         "MmAllocateMappingAddress", // MmAllocateMappingAddress(size, tag)
+        // --- Windows COM allocation
         "CoTaskMemAlloc", // CoTaskMemAlloc(size)
+        // --- Linux kernel memory allocator
         "kmem_alloc", // kmem_alloc(size, flags)
         "kmem_zalloc", // kmem_zalloc(size, flags)
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll]: I don't think these are from Linux. [The Google results for them](https://www.google.com/search?q=kmem_alloc) refer to Solaris and BSD. Where they mention Linux, it's in the context of ZFS or XFS: file systems that were ported to Linux from other kernels and therefore use wrapper functions.

My best suggestion is to write `Solaris/BSD kernel memory allocator` here.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Deallocation.qll [hunk 3]
```diff
@@ -13,29 +13,43 @@ private class StandardDeallocationFunction extends DeallocationFunction {
...
+        // --- Windows Global / Local legacy allocation
+        "LocalFree", "GlobalFree", "LocalReAlloc", "GlobalReAlloc",
+        // --- Windows System Services allocation
+        "VirtualFree",
+        // --- Windows COM allocation
+        "CoTaskMemFree", "CoTaskMemRealloc",
+        // --- Windows Automation
+        "SysFreeString",
+        // --- Linux kernel memory allocator
+        "kmem_free"
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Deallocation.qll:39]: Same as above

**Review by jbj — COMMENTED** [review]
Otherwise LGTM

File: cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll]: Thanks.  Done.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Deallocation.qll (hunk 3 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Deallocation.qll]: Also done.

**Comment by geoffw0** [comment]
Fixed test failure.  It was due to a logic mistake on my part.

**Review by rdmarsh2 — APPROVED** [review]

**Pull request merged**